### PR TITLE
[5.9] Add the drop_deinit instruction

### PR DESCRIPTION
**Pull request by eeckstein**
The `drop_deinit` instruction is a marker for a following destroy instruction to suppress the call of the move-only type's deinitializer.

SILGen needs to insert `drop_deinit` in move-only deinitializers to avoid that - due to other optimizations, like inlining - the deinitializer call is inserted multiple times.

rdar://105798769

This is a cherry-pick of https://github.com/apple/swift/pull/65060

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SIL/IR/SILFunctionType.cpp [hunk 1]
```diff
@@ -2385,7 +2385,7 @@ struct DeallocatorConventions : Conventions {
 
   ParameterConvention
   getIndirectSelfParameter(const AbstractionPattern &type) const override {
-    llvm_unreachable("Deallocators do not have indirect self parameters");
+    return ParameterConvention::Indirect_In;
```
Comment (kavon) [inline lib/SIL/IR/SILFunctionType.cpp:2388]: I'm curious about why this change was needed.

**Review by kavon — APPROVED** [review]
Overall I see no risks in merging this into 5.9 and it looks good. Had a few comments that are mostly clarifications since i missed reviewing this when it landed into `main`

File: lib/IRGen/IRGenSIL.cpp [hunk 2]
```diff
@@ -1222,6 +1222,10 @@ class IRGenSILFunction :
     auto e = getLoweredExplosion(i->getOperand());
     setLoweredExplosion(i, e);
   }
+  void visitDropDeinitInst(DropDeinitInst *i) {
+    auto e = getLoweredExplosion(i->getOperand());
+    setLoweredExplosion(i, e);
+  }
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/IRGen/IRGenSIL.cpp:1228]: True. This could be an unreachable.
I just copied from `visitMoveValueInst`.
But you are right, we should replace both visit functions with unreachable

File: lib/SIL/IR/SILFunctionType.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SIL/IR/SILFunctionType.cpp:2388]: This is for supporting deinits which get the self argument as indirect argument, which is the case for move-only types.
